### PR TITLE
2521 delete subscription resource

### LINKED ISSUE #3234: When subscriptions cannot be activated hapi-fhir fails indefinitely on startup
When a subscription is created that cannot be activated, failures result that end up in an infinite hang loop for hapi-fhir/smile

Issue from SMILE:
After deleting a Subscription resource on 08.R02, on startup it loops 60 times over this error and then fails to start up the subscription module. Reinserting the Subscription resource makes the error go away.
15:47:02.953 [main] ERROR c.u.f.jpa.searchparam.retry.Retrier - Retry failure 9/60: ca.uhn.fhir.rest.server.exceptions.ResourceGoneException: Resource was deleted at 2021-10-15T03:40:37.056+00:00

Steps to Reproduce
PUT the following Subscription resource on the server:
`{
"resourceType": "Subscription",
"id": "32617765",
"status": "off",
"criteria": "Consent?",
"channel": {
"type": "message",
"endpoint": "channel:nalc.prod.sub.consent",
"payload": "application/json"
}
}
After a time, DELETE the Subscription resource. On restart of Smile it goes into the 60x retry loop before it finishes starting up the other modules.
Note that this instance of Smile is using Azure Event Hub as the broker if that is a factor.
`

Observed Results
On startup it loops 60 times over this error and then fails to start up the subscription module.
15:47:02.953 [main] ERROR c.u.f.jpa.searchparam.retry.Retrier - Retry failure 9/60: ca.uhn.fhir.rest.server.exceptions.ResourceGoneException: Resource was deleted at 2021-10-15T03:40:37.056+00:00

### PR DESCRIPTION
**Pull request by TipzCM**
close #3234

Changes:
* Added error handling in SubscriptionActivatingSubscriber to set failed activations to ERROR state
* Added logging to SubscriptionLoader to log when subscriptions fail to activate
* Added changelog

### CODE REVIEW

**Review by jdar8 — COMMENTED** [review]
Looks good! Left a couple comments

File: hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/match/matcher/subscriber/SubscriptionActivatingSubscriber.java [hunk 1]
```diff
@@ -115,15 +116,23 @@ public boolean activateSubscriptionIfRequired(final IBaseResource theSubscriptio
...
+			subscription =  subscriptionDao.read(theSubscription.getIdElement());
+
+			subscription.setId(subscription.getIdElement().toVersionless());
+
+			ourLog.info("Activating subscription {} from status {} to {}", subscription.getIdElement().toUnqualified().getValue(), SubscriptionConstants.REQUESTED_STATUS, SubscriptionConstants.ACTIVE_STATUS);
 			SubscriptionUtil.setStatus(myFhirContext, subscription, SubscriptionConstants.ACTIVE_STATUS);
 			subscriptionDao.update(subscription);
 			return true;
-		} catch (final UnprocessableEntityException e) {
+		} catch (final Exception e) {
```
Comment (tadgh) [inline hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/match/matcher/subscriber/SubscriptionActivatingSubscriber.java:131]: Avoid catching exception where possible, favour catching a specific exception or class of exceptions.

File: hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/match/registry/SubscriptionLoader.java [hunk 2]
```diff
@@ -56,8 +56,8 @@
 
 public class SubscriptionLoader implements IResourceChangeListener {
 	private static final Logger ourLog = LoggerFactory.getLogger(SubscriptionLoader.class);
-	private static final int MAX_RETRIES = 60; // 60 * 5 seconds = 5 minutes
-	private static long REFRESH_INTERVAL = DateUtils.MILLIS_PER_MINUTE;
+	private static final int MAX_RETRIES = 1; // 60 * 5 seconds = 5 minutes
+	private static final long REFRESH_INTERVAL = DateUtils.MILLIS_PER_MINUTE;
```
Comment (tadgh) [inline hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/match/registry/SubscriptionLoader.java:60]: can you move the comment one line lower? I think this comment is on the wrong line.

File: hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/match/registry/SubscriptionLoader.java [hunk 3]
```diff
@@ -111,7 +111,8 @@ public void syncSubscriptions() {
 		}
 		try {
 			doSyncSubscriptionsWithRetry();
-		} finally {
+		}
+		finally {
```
Comment (tadgh) [inline hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/match/registry/SubscriptionLoader.java:115]: Nitpick: don't change the brace style on existing code. fwiw we should have an autoformatter which lints on checkin, but alas.

File: hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/match/registry/SubscriptionLoader.java [hunk 4]
```diff
@@ -130,6 +131,8 @@ public int doSyncSubscriptionsForUnitTest() {
 	}
 
 	synchronized int doSyncSubscriptionsWithRetry() {
+		// retry runs MAX_RETRIES times
+		// and if errors result every time, it will fail
```
Comment (tadgh) [inline hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/match/registry/SubscriptionLoader.java:135]: is this comment still valid or was it just for development purposes?

File: hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/match/registry/SubscriptionLoader.java [hunk 5]
```diff
@@ -182,10 +185,18 @@ private int updateSubscriptionRegistry(List<IBaseResource> theResourceList) {
...
 			boolean activated = mySubscriptionActivatingInterceptor.activateSubscriptionIfRequired(resource);
 			if (activated) {
 				activatedCount++;
 			}
+			else {
+				ourLog.error("Subscription "
+					+ resource.getIdElement().getIdPart()
+					+ " could not be activated."
+					+ " This will not prevent startup, but it could lead to undesirable outcomes!");
+			}
```
Comment (tadgh) [inline hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/match/registry/SubscriptionLoader.java:199]: Can you also read the error information out of the subscription resource and add it to this log line? Would be useful for users.

**Review by tadgh — APPROVED** [review]
Approved pending minor stuff. Hold off on merging this until we try it against azure.

File: hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/match/matcher/subscriber/SubscriptionActivatingSubscriber.java (hunk 1 above)
Comment (TipzCM) [reply hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/match/matcher/subscriber/SubscriptionActivatingSubscriber.java]: updated

File: hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/match/registry/SubscriptionLoader.java (hunk 2 above)
Comment (TipzCM) [reply hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/match/registry/SubscriptionLoader.java:61]: ah. actually, i think i just left debug code in... number should be 60, not 1 (but i didn't want to wait 60 iterations of failed startups... each taking about 5 seconds)

File: hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/match/registry/SubscriptionLoader.java (hunk 4 above)
Comment (TipzCM) [reply hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/match/registry/SubscriptionLoader.java:138]: No, this comment is correct.

I wrote it for future generations

File: hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/match/matcher/subscriber/SubscriptionActivatingSubscriber.java [hunk 6]
```diff
@@ -118,17 +121,24 @@ private boolean activateSubscription(final IBaseResource theSubscription) {
...
+			subscription.setId(subscription.getIdElement().toVersionless());
+
+			ourLog.info("Activating subscription {} from status {} to {}", subscription.getIdElement().toUnqualified().getValue(), SubscriptionConstants.REQUESTED_STATUS, SubscriptionConstants.ACTIVE_STATUS);
 			SubscriptionUtil.setStatus(myFhirContext, subscription, SubscriptionConstants.ACTIVE_STATUS);
 			subscriptionDao.update(subscription, srd);
 			return true;
-		} catch (final UnprocessableEntityException e) {
+		} catch (final UnprocessableEntityException | ResourceGoneException e) {
+			subscription = subscription != null ? subscription : theSubscription;
+			ourLog.info("Failed to activate subscription "
```
Comment (tadgh) [inline hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/match/matcher/subscriber/SubscriptionActivatingSubscriber.java:137]: This should be an error log or at least a warn log

**Review by tadgh — APPROVED** [review]

File: hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/match/matcher/subscriber/SubscriptionActivatingSubscriber.java (hunk 6 above)
Comment (TipzCM) [reply hapi-fhir-jpaserver-subscription/src/main/java/ca/uhn/fhir/jpa/subscription/match/matcher/subscriber/SubscriptionActivatingSubscriber.java]: done

**Pull request merged**